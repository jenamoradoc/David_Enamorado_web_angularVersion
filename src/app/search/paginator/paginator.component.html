<div class="paginator">
  <div class="per-page">
    <label for="per-page">Items per page: </label>
    <select name="per-page" id="per-page" [formControl]="changeSizePageControl" #perPage>
      <option *ngFor="let pageSize of pageSizeOptions" [value]="pageSize">{{pageSize}}</option>
    </select>
  </div>
  <div class="info-page">
    <p>Showing {{ from }}-{{ to }} of {{ totalCount }} results</p>
  </div>
  <div class="paginator-actions">
    <button class="btn btn-paginator" (click)="onPrev()" [disabled]="to === totalCount"> < </button>
    <button class="btn btn-paginator" (click)="onNext()" [disabled]="to === totalCount "> > </button>
  </div>
</div>
