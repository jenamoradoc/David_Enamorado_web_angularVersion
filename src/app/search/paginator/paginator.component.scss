.paginator {
    margin-top: 10px;
    display: flex;
    align-items: center;
    width: 100%;

    .per-page {
        margin-right: 1rem;
        font-size: 0.825rem;
        color: #999;
    }

    .info-page {
        flex: 1;
        text-align: center;
        font-size: 0.825rem;
        color: #999;
    }

    .btn-paginator {
        margin: 0 5px;
        padding: 5px 10px;
        border: 1px solid #ccc;
        border-radius: 5px;
        transition: all 0.3s ease-in-out;
        font-weight: 900;
        color: #999;

        &:hover {
            background-color: #ccc;
        }
    }
}