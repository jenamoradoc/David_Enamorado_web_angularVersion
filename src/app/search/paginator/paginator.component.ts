import { outputAst } from '@angular/compiler';
import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { FormControl } from '@angular/forms';
import { distinctUntilChanged, startWith, tap } from 'rxjs/operators';

@Component({
  selector: 'app-paginator',
  templateUrl: './paginator.component.html',
  styleUrls: ['./paginator.component.scss'],
})
export class PaginatorComponent {
  changeSizePageControl = new FormControl<number>(10, { nonNullable: true });

  @Input() pageSizeOptions = [10, 20, 50, 100];
  @Input() totalCount = 0;
  @Output() onChangePage = new EventEmitter<number>();
  @Output() onChangePageSize = this.changeSizePageControl.valueChanges.pipe(
    distinctUntilChanged(),
  );
  // reset paginator
  private _page = 1;
  @Input() set page(value: number) {
    this._page = value;
  }

  get page() {
    return this._page;
  }

  get from() {
    return (this._page - 1) * this.changeSizePageControl.value + 1;
  }

  get to() {
    return Math.min(
      this._page * this.changeSizePageControl.value,
      this.totalCount
    );
  }

  onPrev() {
    if (this._page > 1) {
      this._page -= 1;
      this.onChangePage.emit(this._page);
    }
  }

  onNext() {
    if (this._page < this.changeSizePageControl.value) {
      this._page += 1;
      this.onChangePage.emit(this._page);
    }
  }
}
