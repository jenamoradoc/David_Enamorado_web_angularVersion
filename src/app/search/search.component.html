<div class="container">
  <div class="search-form">
    <!-- Search form -->
    <form [formGroup]="searchForm" (ngSubmit)="onSubmit()" class="form-group">
      <div class="form-control">
        <label for="search">Search Github Users</label>
        <input
          type="text"
          class="form-input"
          id="search"
          placeholder="Type an username..."
          formControlName="search"
        />
        <div class="actions">
            <button type="submit" class="btn btn-primary" style="margin-right: 0.25rem;" [disabled]="searchForm.invalid">Submit</button>
            <button type="reset" class="btn btn-secondary" (click)="onClear(); $event.preventDefault()">Clear</button>
        </div>
      </div>
    </form>
  </div>
  <div class="users-container">
    <ng-container *ngIf="showSplash; else showUsers">
      <div class="splash">
        <h2>Search Github Users</h2>
        <p>Enter a username to search for a Github user</p>
        <img src="assets/undraw_team_re_0bfe.svg" alt="github_users_splash">
      </div>
    </ng-container>
    <ng-template #showUsers>
      <div class="sort-select" [formGroup]="searchForm">
        <select name="sort" id="sort" formControlName="sort">
          <option value="best match">Best Match</option>
          <option value="followers">Followers</option>
          <option value="repositories">Repositories</option>
          <option value="joined">Joined</option>
        </select>
      </div>
      <app-users [users]="users"></app-users>
      <app-paginator [page]="currentPage" [totalCount]="total_count" (onChangePage)="onChangePage($event)" (onChangePageSize)="onChangePageSize($event)"></app-paginator>
    </ng-template>
  </div>
</div>
