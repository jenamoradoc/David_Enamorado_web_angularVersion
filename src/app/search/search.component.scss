.container {
  display: grid;
  align-items: center;
  grid-template-columns: 2fr 3fr;

  .search-form {
    background-image: linear-gradient(0deg, #accbee 0, #e7f0fd);
    border-right: 1px solid #d8d8d8;
    display: flex;
    height: 100vh;
    justify-content: center;
    padding: 0 1.25rem;
    position: relative;

    .form-group {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      width: 100%;
    }

    .form-control {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      width: 100%;
      padding: 1rem 2rem;
      box-sizing: border-box;
      background-color: #fff;
      border-radius: 5px;
      box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);

      label {
        font-size: 1.5rem;
        font-weight: 500;
        margin-bottom: 1rem;
      }

      .form-input {
        width: 100%;
        height: 40px;
        border: 1px solid #ccc;
        border-radius: 5px;
        padding: 0 10px;
        box-sizing: border-box;
        margin-bottom: 10px;
      }
    }

    .search input {
      width: 100%;
      height: 40px;
      border: 1px solid #ccc;
      border-radius: 4px;
      padding: 12px 20px;
      margin: 8px 0;
      box-sizing: border-box;
    }
  }
  .users-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-around;
    overflow-y: hidden;
    position: relative;
    height: 100vh;
    padding: 1rem;

    .sort-select {
      margin-bottom: 10px;
      select {
        height: 40px;
        border: 1px solid #ccc;
        border-radius: 4px;
        padding: 12px 20px;
        // margin: 8px 0;
        box-sizing: border-box;
      }
    }

    app-users {
      overflow-y: auto;
    }
    .splash {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      p {
        margin-bottom: 1rem;
      }
      img {
        width: 20rem;
      }
    }
  }
}
