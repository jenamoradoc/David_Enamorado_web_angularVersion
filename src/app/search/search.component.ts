import { Component, OnInit } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { switchMap } from 'rxjs/operators';
import { GithubService } from '../github/github.service';
import { SearchForm } from '../types/search.types';
import { Item } from '../types/user.types';

@Component({
  selector: 'app-search',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.scss'],
})
export class SearchComponent implements OnInit {
  showSplash = true;
  searchForm = new FormGroup<SearchForm>({
    search: new FormControl<string>('', {
      nonNullable: true,
      validators: [Validators.required],
    }),
    page: new FormControl<number>(1, { nonNullable: true }),
    per_page: new FormControl<number>(10, { nonNullable: true }),
    sort: new FormControl<string>('best match', { nonNullable: true }),
  });
  sortControl = new FormControl<string>('best match', { nonNullable: true });

  users: Item[] = [];
  total_count = 0;
  currentPage = 1;

  constructor(private _githubService: GithubService) {}

  ngOnInit(): void {
    this.searchForm.get('sort')?.valueChanges.pipe(
      switchMap((sort) => {
        const searchTerms = this.searchForm.getRawValue();
        return this._githubService.searchUsers({...searchTerms, sort, page: this.currentPage});
      }),
    ).subscribe(({ items, total_count }) => {
      this.users = items;
      this.total_count = total_count;
    });
  }

  onSubmit(): void {
    this.showSplash = false;
    this.currentPage = 1;
    this.searchUsers();
  }

  searchUsers(): void {
    const searchTerms = this.searchForm.getRawValue();
    this._githubService.searchUsers({...searchTerms, page: this.currentPage}).subscribe(({ items, total_count }) => {
      this.users = items;
      this.total_count = total_count;
    });
  }

  onClear(): void {
    this.showSplash = true;
    this.searchForm.reset({
      search: '',
      page: 1,
      per_page: 10,
      sort: 'best match',
    }, { emitEvent: false });
  }

  onChangePage(page: number): void {
    this.currentPage = page;
    this.searchForm.patchValue({ page });
    this.searchUsers();
  }

  onChangePageSize(per_page: number): void {
    this.currentPage = 1;
    this.searchForm.patchValue({ per_page });
    this.searchUsers();
  }
}
