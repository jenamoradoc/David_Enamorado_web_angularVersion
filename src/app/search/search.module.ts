import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { SearchRoutingModule } from './search-routing.module';
import { SearchComponent } from './search.component';
import { ReactiveFormsModule } from '@angular/forms';
import { UsersComponent } from './users/users.component';
import { PaginatorComponent } from './paginator/paginator.component';
import { UserComponent } from './user/user.component';


@NgModule({
  declarations: [
    SearchComponent,
    UsersComponent,
    PaginatorComponent,
    UserComponent
  ],
  imports: [
    CommonModule,
    SearchRoutingModule,
    ReactiveFormsModule,
  ]
})
export class SearchModule { }
