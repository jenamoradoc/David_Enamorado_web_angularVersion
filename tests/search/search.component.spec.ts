import { ComponentFixture, TestBed } from '@angular/core/testing';
import { HttpClientTestingModule } from '@angular/common/http/testing';

import { SearchComponent } from '../../src/app/search/search.component';
import { ReactiveFormsModule } from '@angular/forms';

import { PaginatorComponent } from '../../src/app/search/paginator/paginator.component';

describe('SearchComponent', () => {
  let component: SearchComponent;
  let fixture: ComponentFixture<SearchComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [SearchComponent, PaginatorComponent],
      imports: [HttpClientTestingModule, ReactiveFormsModule],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(SearchComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should search users on submit', () => {
    component.onSubmit();
    expect(component.showSplash).toBeFalsy();
  });
});
