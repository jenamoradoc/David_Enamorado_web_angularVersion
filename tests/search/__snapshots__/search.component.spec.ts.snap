// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="container"
  >
    <div
      class="search-form"
    >
      <form
        class="form-group ng-untouched ng-pristine ng-invalid"
        novalidate=""
      >
        <div
          class="form-control"
        >
          <label
            for="search"
          >
            Search Github Users
          </label>
          <input
            class="form-input ng-untouched ng-pristine ng-invalid"
            formcontrolname="search"
            id="search"
            placeholder="Type an username..."
            type="text"
          />
          <div
            class="actions"
          >
            <button
              class="btn btn-primary"
              disabled=""
              style="margin-right: 0.25rem;"
              type="submit"
            >
              Submit
            </button>
            <button
              class="btn btn-secondary"
              type="reset"
            >
              Clear
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      class="users-container"
    >
      <div
        class="splash"
      >
        <h2>
          Search Github Users
        </h2>
        <p>
          Enter a username to search for a Github user
        </p>
        <img
          alt="github_users_splash"
          src="assets/undraw_team_re_0bfe.svg"
        />
      </div>
      
      
      
    </div>
  </div>
</div>
`;
